src/odb/include/odb/db.h

```diff
@@ -7617,22 +7639,91 @@ class dbPowerSwitch : public dbObject
 class dbScanChain : public dbObject
 {
  public:
+  dbSet<dbScanPartition> getScanPartitions() const;
+
+  dbSet<dbScanInst> getScanInsts() const;
+
+  // User Code Begin dbScanChain
+  std::string_view getName() const;
+
+  void setName(std::string_view name);
+
+  void setScanIn(dbBTerm* scan_in);
+  dbBTerm* getScanIn() const;
+
+  void setScanOut(dbBTerm* scan_out);
+  dbBTerm* getScanOut() const;
+
+  void setScanEnable(dbBTerm* scan_enable);
+
+  dbBTerm* getScanEnable() const;
+
+  std::string_view getTestMode() const;
+  void setTestMode(std::string_view test_mode);
+
+  dbScanPartition* createScanPartition();
+  dbScanInst* createScanInst(dbInst* inst);
+  dbScanInst* createScanInst(const std::vector<dbInst*>& insts);
+  // User Code End dbScanChain
 };
 
 class dbScanInst : public dbObject
 {
  public:
-  enum class SCAN_INST_TYPE
+  struct AccessPins
   {
-    OneBit,
-    ShiftRegister,
-    BlackBox
+    std::variant<dbBTerm*, dbITerm*> scan_in;
+    std::variant<dbBTerm*, dbITerm*> scan_out;
   };
+  enum class ClockEdge
+  {
+    Rising,
+    Falling
+  };
+
+  // User Code Begin dbScanInst
+  void setScanClock(std::string_view scan_clock);
+  std::string_view getScanClock() const;
+
+  void setClockEdge(ClockEdge clock_edge);
+  ClockEdge getClockEdge() const;
+
+  void setBits(uint bits);
+
+  uint getBits() const;
+
+  void setScanEnable(dbBTerm* scan_enable);
+  void setScanEnable(dbITerm* scan_enable);
+  std::variant<dbBTerm*, dbITerm*> getScanEnable() const;
+
+  void setAccessPins(const AccessPins& access_pins);
+
+  AccessPins getAccessPins() const;
+
+  std::vector<dbInst*> getInsts() const;
```

Comment:
The logic is to be able to express the DEF file section for scan chains for FLOATING and ORDERED elements.

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/1619948/40aabc9c-9b84-4b75-be71-9617bdcb72c1)

FLOATING is easy, it is just one element in each scan inst. For ORDERED, we need to keep the order of the scan elements so we abstract the access pins and we treat that sequence of elements as one "virtual" cell.

So, we could have something like this:

[FF1, FF2, [FF3, FF4], FF5]

scanInst 1: FF1 -> FLOATING
scanInst 2: FF2 -> FLOATING
scanInst 3: [FF3, FF4] -> ORDERED
scanInst 4: FF5 -> FLOATING

One example of ORDERED elements are shift-registers that you are able to use it for shifting patterns but they should not be reorder (because you will break the functional part of your design).

Reply:
dbScanInst is an odd name to express this as this seems to be an ordered collection of instances.  It feels like there is a missing intermediate class, something like

dbScanChain -> dbScanInstSequence -> dbScanInst

Reply:
That makes sense, what do you think if I reuse the name ScanCell here:

dbScanChain -> dbScanCell -> dbScanInst

Similar to the ScanCell in src/dft/src/cells/ScanCell.hh where we hold dbInsts ?

Reply:
Instance and Cell sound to me like placeable objects rather than abstract collections.  Do you have thoughts on how we might avoid terminology conflict?

Reply:
My concern with (2) is can you get it through swig into tcl/python.  If you want to see if that's possible then I'm fine with (2).

Reply:
That's fair. I always forget about tcl. I will go with (1) creating dbScanList to hold a sequence of dbScanInst then.

Thanks!